src/test/fuzz/bech32.cpp

```diff
@@ -16,28 +16,27 @@
 FUZZ_TARGET(bech32)
 {
     const std::string random_string(buffer.begin(), buffer.end());
-    const std::pair<std::string, std::vector<uint8_t>> r1 = bech32::Decode(random_string);
-    if (r1.first.empty()) {
-        assert(r1.second.empty());
+    const auto r1 = bech32::Decode(random_string);
+    if (r1.hrp.empty()) {
+        assert(r1.data.empty());
     } else {
-        const std::string& hrp = r1.first;
-        const std::vector<uint8_t>& data = r1.second;
-        const std::string reencoded = bech32::Encode(hrp, data);
+        const std::string reencoded = bech32::Encode(r1.encoding, r1.hrp, r1.data);
         assert(CaseInsensitiveEqual(random_string, reencoded));
     }
 
     std::vector<unsigned char> input;
     ConvertBits<8, 5, true>([&](unsigned char c) { input.push_back(c); }, buffer.begin(), buffer.end());
-    const std::string encoded = bech32::Encode("bc", input);
-    assert(!encoded.empty());
 
-    const std::pair<std::string, std::vector<uint8_t>> r2 = bech32::Decode(encoded);
-    if (r2.first.empty()) {
-        assert(r2.second.empty());
-    } else {
-        const std::string& hrp = r2.first;
-        const std::vector<uint8_t>& data = r2.second;
-        assert(hrp == "bc");
-        assert(data == input);
+    for (auto encoding : {bech32::Encoding::BECH32, bech32::Encoding::BECH32M}) {
+        const std::string encoded = bech32::Encode(encoding, "bc", input);
+        assert(!encoded.empty());
+        const auto r2 = bech32::Decode(encoded);
+        if (r2.hrp.empty()) {
```

Comment:
Should a round trip encode/decode always result in the same hrp ("bc") here? Am I missing some way that this could legitimately fail?

Reply:
This already existed in the code before this PR, but it's a good point. The only failure case is when the data is too big to fit in a valid Bech32(m) encoding (max 90 characters).

I've added a check for that, and then dropped the `if (r2.hrp.empty()) {` branch.